File: 55.jump-game.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

public class Solution55
{
    // iterate the array from left
    // tryint to use a number with maxreach to tell the current max reach position
    // Break the iteration if the reach can not reach to the current point.
    public bool CanJump(int[] nums) 
    {
        int size = nums.Length;
        if(size == 0) return true;
        int reach = 0;
        for(int i = 0; i < size; ++i)
        {
            if(reach < i || reach >= size -1) break;

            reach = Math.Max(reach, i + nums[i]);
        }

        return reach >= size-1;
    }

    // Start from last position and 
    public bool CanJump2(int[] nums) {
        int lastPos = nums.Length - 1;
        for (int i = nums.Length - 1; i >= 0; i--) {
            if (i + nums[i] >= lastPos) {
                lastPos = i;
            }
        }
        return lastPos == 0;
    }
}

File: 55_jump-game.py
from typing import List


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        size = len(nums)
        reach = 0
        if size == 0: return True
        for i in range(size):
            if i > reach or reach >= size -1: break
            reach = max(reach, i + nums[i])
        
        return reach >= size -1

    def canJump2(self, nums: List[int]) -> bool:
        next = [0]

        while len(next) > 0:
            idx = next.pop(0)
            if idx >= len(nums) - 1:
                print(idx) 
                return True
            if nums[idx] < 1: continue 
            for i in range(1, nums[idx] + 1):
                next.append(idx + i)
            
            print(next)
            

        return False

instance = Solution()
# print(instance.canJump([3,2,1,0,4]))
print(instance.canJump([2,3,1,1,4]))



